### DISCUSSION TITLE
What does `azd env refresh` actually do?

### DISCUSSION THREAD

**Discussion by TWolversonReply** [Azure/azure-dev#4024] [Q&A]
The [documentation](https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/reference#azd-env-refresh) says

> Refresh environment settings by using information from a previous infrastructure provision.

This is rather like defining a term using itself, so could the documentation please specify
- what is meant by 'refresh'
- what information, retrieved how?

I've recently dealt with a question from a team member over some unexpected behaviour from this command, and my suspicion is that there is a bad assumption on the part of azd that only one azd-enabled project will deploy to a subscription. This is not the case for us, we are using several azd projects in this subscription, and if it just takes the most recent deployment, this is likely to be randomly wrong for our usage.

**Comment by vhvb1989** [comment]
Hello @TWolversonReply 

By `refresh` it means update the values from the `.azure/<env name>/.env` file. 

**little context:**
  azd automatically saves the outputs from your infrastructure in the .env file. For example, if you are using `bicep` and define outputs there like:

```bicep
output OUTPUT_NAME string = something.here
```

then, after `azd provision` finishes successfully, azd will save `OUTPUT_NAME=<value>` in the .env
At the same time, all the outputs are persisted as part of the deployment operation in Azure, allowing azd to read/pull this outputs. Azd just need the `name of the environment` you picked, the `subscription` and the `location`.  Having those 3 values, azd can find the last deployment for that project and pull the outputs directly into the env w/o creating a new deployment.

This operation is useful for some cases like:
- You run `azd provision` in one device and you want to run `azd deploy` from a different device. Then, you just need to share the environment name, subscription and location to run `azd refresh` from the second device before running `azd deploy`

You can have multiple azd projects deployed in the same Azure subscription. Even one single project deployed multiple times in the same subscription by picking a new azd-environment name for each one of them. Azd will fetch the state (last successful deployment outputs) based on the azd-environment name+sub+location

**Reply by TWolversonReply** [reply]
Thanks. The point I was making was really, "can this be documented?"

The last point - it's not intuitive that an environment name is expected to be different per project in my opinion. Most people would consider an "environment" to be something like 'dev'. That would be the same across projects - why wouldn't it be? - and we've run with this assumption, which would be annoying to change at this point.

**Reply by rajeshkamal5050** [reply]
Tracking documentation issue - https://github.com/Azure/azure-dev/issues/4368

**Reply by webstean** [reply]
Agreed - the doco does seem to have improvement around this: https://learn.microsoft.com/en-us/azure/developer/azure-developer-cli/environment-variables-faq
